bin/runtime-common/src/messages_call_ext.rs

```diff
@@ -21,18 +21,24 @@ use bp_messages::{LaneId, MessageNonce};
 use frame_support::{dispatch::CallableCallFor, traits::IsSubType, RuntimeDebug};
 use pallet_bridge_messages::{Config, Pallet};
 use sp_runtime::transaction_validity::TransactionValidity;
+use sp_std::ops::RangeInclusive;
 
 /// Generic info about a messages delivery/confirmation proof.
 #[derive(PartialEq, RuntimeDebug)]
 pub struct BaseMessagesProofInfo {
+	/// Message lane, used by the call.
 	pub lane_id: LaneId,
-	pub best_bundled_nonce: MessageNonce,
+	/// Nonces of messages, included in the call.
+	pub bundled_range: RangeInclusive<MessageNonce>,
+	/// Nonce of the best message, stored by this chain before the call is dispatched.
 	pub best_stored_nonce: MessageNonce,
 }
 
 impl BaseMessagesProofInfo {
+	/// Returns true if `bundled_range` cannot be directly appended to the `best_stored_nonce`
+	/// or if the `bundled_range` is empty.
 	fn is_obsolete(&self) -> bool {
-		self.best_bundled_nonce <= self.best_stored_nonce
+		*self.bundled_range.start() != self.best_stored_nonce + 1 || self.bundled_range.is_empty()
```

Comment:
this may be dumb, I haven't followed **all** path, but afaict we allow `FromBridgedChainMessagesProof` with `nonces_end < nonces_start` => here we might "wrongly" report `is_obsolete == false` - which may also be fine in that case since no message will be included ; not sure, just wanted to bring it to your attention

Reply:
You're right, thanks for spotting that. I've changed code a bit. Now we allow empty message delivery transactions but only if there's no room left in the unrewaded relayers vector. So we allow only when lane is blocked. Honest relayers don't want to block the lane, so they'll include reward confirmation in regular message delivery transactions.

Reply:
Actually I think it needs to be implemented in other way: We need to reject delivery transaction with the message that occupies the last free entry in unrewarded relayers vector if this transaction doesn't include rewards confirmation. In current impl, other relayer may submit empty delivery tx with larger tip and the other relayer won't be refunded for the same information (lose his funds).

Reply:
> In current impl, other relayer may submit empty delivery tx with larger tip and the other relayer won't be refunded for the same information (lose his funds).

This is wrong - the second transaction will be considered stale, because `is_obsolete` would return `true` in this way, so it won't be included into the block => no lose-his-funds. OTOH if we'll forbid empty delivery transactions, we'll allow single relayer to block the lane (that's the reason why we have such transactions). So current code should be fine. PTAL when you have a time @acatangiu

Reply:
Filed #2026 to fix relayer code if this PR is accepted

Reply:
Another fix for this PR: right now it would accept empty delivery transactions when there's no empty entries in the unrewarded relayers vec. But there's another constraint, which limits total number of unconfirmed messages (e.g. there may be 1 relayer that has delivered 4096 messages and the lane is blocked until rewards are confirmed. So the empty-delivery-transaction allowance condition needs to be fixed

Reply:
Fixed, please take a look